### ISSUE TITLE
Menu: "Parent link" options only ever shows current language and neutral options

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#4857]
**Description of the bug**

When I am adding a menu link for the Spanish language (and url contains `langcode=es`), I usually want to put the new link underneath another link that is Spanish. The select list for "Parent link" does not show any links in Spanish.

**Steps To Reproduce**

To reproduce the behavior:
1. Enable content translation and local modules
1. Add a second language, for example, Spanish
1. Go to 'Manage links' for the Primary Navigation
1. Click on 'Add link' Select the language "Spanish", 
	* And add the link
1. Click on 'Add link again' Select the language "Spanish", 
	* Scroll down to 'Parent link' option, confirm there are no Spanish links there

**Actual behavior**

Shows only english links for all menus

**Expected behavior**

Show all links in this menu

**Comment by indigoxela** [comment]
Confirmed (sort of).

The new menu item langcode gets set from the url parameter, but the available parent items get set from the language prefix (or domain).

With this in the url: `/de/admin/structure/menu/manage/main-menu/add?langcode=en`

The menu item langcode gets set to "en", but the available parents are "de" and "und".

With this in the url: `/en/admin/structure/menu/manage/main-menu/add?langcode=de`

The menu item langcode gets set to "de", but the available parents are "en" and "und".

This seems inconsistent. Only if *both* values are the same (language prefix and langcode parameter), it currently makes sense.

**Comment by indigoxela** [comment]
Ouch, it's actually even worse...

If editing an existing menu item, which has already a langcode set, but one edits it in another language, the only parent items available in the dropdown list are in the current language, hence the wrong ones.

For instance on /en/admin/structure/menu/item/362/edit, when item 364 is a German menu item, the dropdown list contains only English and neutral items.

Another example: Even if everything is fine (one edits a menu item in the language of the item), but one switches the language of the item, the dropdown list doesn't update, so the available parent items are the wrong ones then... Probably not solvable without ajax.

**Comment by indigoxela** [comment]
A little more details to get that inconsistency, when *adding* new menu items:

On admin/structure/menu/manage/main-menu you can filter the displayed menu items by language. This also sets the url of the "Add link".

If you filter to only display English items, but the current language is another one, then you click on the link to add a new item, the urls get: `/de/admin/structure/menu/manage/main-menu/add?langcode=en`
The langcode of the new item gets set to "English", but in the "Parent link" dropdown you only get English and neutral items as options.

**Comment by indigoxela** [comment]
A [PR](https://github.com/backdrop/backdrop/pull/3481) (at an early stage) is available for testing.

This should at least improve the situation. Note, there's no dynamic switch if the langcode gets changed.

To my understanding it's function menu_tree_all_data() that needs an additional parameter to make language selection available for function _menu_get_options().

Currently the language is hardcoded in [menu_tree_all_data()](https://api.backdropcms.org/api/backdrop/core%21includes%21menu.inc/function/menu_tree_all_data/1)

**Comment by indigoxela** [comment]
How to test:

1. Enable Content Translation, Language, Locale modules
2. Add some languages (admin/config/regional/language/add)
3. Add menu items to a menu of your choice
4. Verify that the "Parent link" dropdown list always contains the right ones (same language or language neutral, all if the menu item is "language neutral")
5. Do the same with existing menu items (e.g. switch language)

**Comment by herbdool** [comment]
I think this is a duplicate issue. I need to look it up though.

**Comment by herbdool** [comment]
Maybe it's not a duplicate. I mentioned it on https://github.com/backdrop/backdrop-issues/issues/4774 but it's not the same.

**Comment by herbdool** [comment]
And we'll need to ensure it doesn't conflict with https://github.com/backdrop/backdrop-issues/issues/4769.

**Comment by indigoxela** [comment]
@herbdool I'm confident, that the taxonomy term issue is unrelated.

I'll have a closer look at the node form issue.

**Comment by herbdool** [comment]
@indigoxela approach works. The Parent link is filtered by "Show menu in language" rather than the interface language. Though it seems that some people prefer it to show all menu links with no filtering. I'm not sure which is best. But if it did show all links regardless of language then it would make it easier to deal with the Parent item on the node edit form.

Currently the Menu Parent item on the node edit form filters by the interface. But with this PR it changes the behaviour to showing all links regardless of language. So seems that we need to address both in this PR so the experience is consistent.

**Comment by jenlampton** [comment]
The menu UI defaulted to all links unless I specifically filtered by language. 👍 
The node edit form defaulted to Spanish on a Spanish translation, and English on an English translation 👍 

Unfortunately, when creating a new translatable node, it only showed the English parents (site default language / interface language) regardless of the language selected on the node form. If I were creating a Spanish node, I wouldn't be able to add it into the Spanish menu. I think that interface should show all menu links instead of a single language. 

I may have identified what is causing this in the PR. I left a note there.

**Comment by indigoxela** [comment]
I think, we need some clarification here, it seems to me there's some confusion about what my PR tries to achieve.

My PR does not (yet) fix the _node form_ parent dropdown. Mainly to not conflict with issue 4769, but also because the node form brings more potential UX problems. See this [issue comment](https://github.com/backdrop/backdrop-issues/issues/4769#issuecomment-755935427) with my concerns.

My PR really only fixes the parent select list, when adding new or editing existing menu items via the menu admin form (menu_edit_item). It uses Ajax to dynamically update the parent select list.

What my PR does, is to filter the parent list for usable ones - items that are either in the same language or language neutral, to reduce the danger of creating impossible combinations. A Spanish item with an English parent wouldn't ever be visible for visitors, so we shouldn't offer it in the options.

I'd not recommend to show all items (regardless of language) by default. That might work for sites with few menu items and with no more than maybe three languages. But think of a site with 25 nested menu items and 6 languages... how would anyone be able to figure out the right ones? Keep in mind that there's also a possibility that menu item titles are the same in different languages. Which one would you pick up then? That's like gambling.

@jenlampton many thanks for having a look, but I think the logic is correct as it currently is. See my answer to your comment.

**Comment by indigoxela** [comment]
So, how do we proceed here? Should we also consider the menu item parent item _in the node form_ here, together with the menu admin form? Or should that be a follow-up issue?

@jenlampton and @herbdool what do you think?

**Comment by herbdool** [comment]
Maybe leave the node form for a separate issue. You make a good point about not showing them all.

**Comment by indigoxela** [comment]
> Maybe leave the node form for a separate issue.

On the other hand that issue would be blocked by this one. So it seems to me, it makes sense to fix both problems at once (in node form and in menu item form).

@jenlampton I considered your request to always provide all items in the select list, regardless of their language, but I think that would result in worse UX. See my concerns in my [previous comment](https://github.com/backdrop/backdrop-issues/issues/4857#issuecomment-756601763).

**How to test the updated PR:**

The menu admin form:

- Enable Content translation, Language, Locale modules
- Add some languages
- Go to admin/structure/menu/manage/main-menu and filter by a language
- Create some menu items in different languages "Add link"
- To verify that you always get the right parents, switch interface language and add/edit items
- Change the langcode of existing items, verify that the parent list fits

The node form:

- Set node types to translatable (admin/structure/types/manage/page)
- Add nodes of these types in different languages
- Add menu items via node form
- Verify that the correct items are available as parents
- Switch the language of nodes and verify that the parent list got updated

**Comment by herbdool** [comment]
@indigoxela this is an elegant solution - I like it.

I've tested and reviewed the code. That all looks good. Just need an automated test for this I think and we're good to go.

**Comment by indigoxela** [comment]
> I've tested and reviewed the code.

@herbdool cool, many thanks! The labels seem a bit confusing now, though. I'll add "needs code review" back.

Functional tests have been added.

**Comment by herbdool** [comment]
I've reviewed again. Thanks for the tests and good to go.

Change record can happen after the merge I'm guessing.

**Comment by klonos** [comment]
> Change record can happen after the merge I'm guessing.

Yup 👍

**Comment by jenlampton** [comment]
> So, how do we proceed here? Should we also consider the menu item parent item in the node form here, together with the menu admin form? Or should that be a follow-up issue?

I think a follow-up issue is a good place to address the node form. 

Here's a go at the change record:
https://api.backdropcms.org/change-records/new-parameter-langcode-added-to-menu_tree_all_data

**Comment by quicksketch** [comment]
Thanks folks! I published the change log entry: https://api.backdropcms.org/change-records/new-parameter-langcode-added-to-menu_tree_all_data

I made one last change to add a `@since 1.17.5 $langcode parameter added.` to the docblock of the function.

And I merged https://github.com/backdrop/backdrop/pull/3481 into 1.x and 1.17.x. Thanks @indigoxela for the most excellent work in solving this problem, along with @herbdool, @jenlampton, and @klonos for their work on testing.
